expr/functions/averageSeriesWithWildcards/function.go

```diff
@@ -71,7 +71,8 @@ func (f *averageSeriesWithWildcards) Do(ctx context.Context, e parser.Expr, from
 	for i, series := range nodeList {
 		args := groups[series]
 		name := "averageSeriesWithWildcards(" + series + ")"
-		r := args[0].CopyTag(name, map[string]string{"name": series})
+		r := args[0].CopyTag(name, map[string]string{"name": name})
```

Comment:
Name tag must be cleaned from function artifacts. If not, alias functions need extract metric name from them. Thats really needed ?

Reply:
That's interesting, since it would explain the fix we had to do related to `ExtractMetric`. But we've been taking GraphiteWeb as a reference, and it does seem to include functions in the `name` tag; for example:

https://github.com/graphite-project/graphite-web/blob/ed08fa4e2f6526ae11e55f88f70d0b6d36e4052f/webapp/graphite/render/functions.py#L243-L249

And from what I see, functions like `aliasByNode` are indeed extracting the name from them: https://github.com/graphite-project/graphite-web/blob/ed08fa4e2f6526ae11e55f88f70d0b6d36e4052f/webapp/graphite/render/functions.py#L111

Please let us know if you think we are misunderstanding something.

Reply:
It's not a simple question. ExtractName can extract name/tags from them. But why it's really needed ?
And how to resolve aliasByTags(series, 'name)

With this PR, returned ```averageSeriesWithWildcards... ``` And need some additional strip. May be do it in place, as are in head ? 100% compability of internal structs usually not required. What are changed this for display in graphana dashboard ? But it's disputable.

Reply:
>  100% compability of internal structs usually not required. What are changed this for display in graphana dashboard ? 

Well, the `"name"` tag is not something internal, so setting it to `foobar` vs setting it to `averageSeriesWithWildcards(foobar)` does make a difference. And I think it's not only about visualization, it can lead to different results when combined with other functions; for example  `grep(aliasByNode(averageSeries(...), "name"), "average")`.

Do you have a reference to GraphiteWeb's docs or code that suggest that the `"name"` tag should always be set to the metric name without function names?

Reply:
aliasByNodes/aliasByTags functions will work corect with this, ExtractMetric is used. But may be in other.
https://github.com/grafana/carbonapi/blob/d511d4c9150c851de8e77cd4f3014071ad754e61/expr/types/types.go#L468

But you are right. Some functions like can produce a problem.
Like `grep(aliasByNode(averageSeries(...), "name"), "average").`
It's disputable question, in my opinion search on name without function is better.
But if in graphite-web realize different logic -  so, let it be. No objections.
May be more tests for correct tags extraction is needed, but now I can't say in which functions pipeline it's needed.

PS. Some functions in graphite API is dangerous, like `aliasSub`, which applied to raw name (with functions and args). and can corrupt name.
But capability :-(

Reply:
In that case I think solution is to create internal tag (e.x. `__name__` or something like that) that wouldn't be visible to the user or just internal field in MetricData struct. As one of the problems here is if you have full UTF-8 support enabled and all symbols are allowed in a metric name, you will have problems parsing the result, and while on the parser level you have an idea what is metric name and what is not, but on a function level ExtractMetric will produce malformed results sometimes because at that level we don't have the whole syntax tree, but rather just a string with text.

Reply:
I like the idea of the internal field. That would make the metric name extraction less error-prone and more efficient. For example, this call wouldn't be necessary: https://github.com/go-graphite/carbonapi/pull/686#discussion_r965554117 

But I think it would be better to leave this change out of this PR, I imagine it will be rather big. And I don't see any bugs related to this with the current state of the code anyway, so this PR shouldn't leave the code broken.

If you agree, I can create an issue so we can track it after this is merged.

Reply:
I would agree with that.